ic-transport-types/src/lib.rs

```diff
@@ -122,9 +122,73 @@ pub enum QueryResponse {
     Replied {
         /// The reply from the canister.
         reply: ReplyResponse,
+
+        /// The list of node signatures.
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        signatures: Vec<NodeSignature>,
     },
     /// The request was rejected.
-    Rejected(RejectResponse),
+    Rejected {
+        /// The rejection from the canister.
+        #[serde(flatten)]
+        reject: RejectResponse,
+
+        /// The list of node signatures.
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        signatures: Vec<NodeSignature>,
+    },
+}
+
+impl QueryResponse {
+    /// Returns the signable form of the query response, as described in
+    /// [the spec](https://internetcomputer.org/docs/current/references/ic-interface-spec#http-query).
+    /// This is what is signed in the `signatures` fields.
+    pub fn signable(&self, request_id: RequestId, timestamp: u64) -> Vec<u8> {
+        #[derive(Serialize)]
+        #[serde(tag = "status", rename_all = "snake_case")]
+        enum QueryResponseSignable<'a> {
+            Replied {
+                #[serde(with = "serde_bytes")]
+                reply: &'a Vec<u8>,
+                request_id: RequestId,
+                timestamp: u64,
+            },
+            Rejected {
+                reject_code: RejectCode,
+                reject_message: &'a String,
+                #[serde(default)]
+                error_code: Option<&'a String>,
+                request_id: RequestId,
+                timestamp: u64,
+            },
+        }
+        let response = match self {
+            Self::Replied { reply, .. } => QueryResponseSignable::Replied {
+                reply: &reply.arg,
+                request_id,
+                timestamp,
+            },
+            Self::Rejected { reject, .. } => QueryResponseSignable::Rejected {
+                error_code: reject.error_code.as_ref(),
+                reject_code: reject.reject_code,
+                reject_message: &reject.reject_message,
+                request_id,
+                timestamp,
+            },
+        };
+        let mut signable = Vec::with_capacity(44);
+        signable.extend_from_slice(b"\x0Bic-response");
+        signable.extend_from_slice(to_request_id(&response).unwrap().as_slice());
```

Comment:
to_request_id could panic, should an agent error be returned instead? 
Now, a malicious replica can crash the software using agent-rs.

One way this might panic is if the RequestIdSerializer is handed a QueryResponseSignable::Rejected with an error code outside the range 1-5.

Reply:
Where are you seeing that this would cause a panic?